File: src/test/java/GeneralTest.java
import de.floribe2000.warships_java.direct.account.PlayersPersonalDataFull;
import de.floribe2000.warships_java.direct.account.PlayersPersonalDataFullRequest;
import de.floribe2000.warships_java.direct.api.ApiBuilder;
import de.floribe2000.warships_java.direct.api.typeDefinitions.Region;
import de.floribe2000.warships_java.direct.general.ServerStatus;
import de.floribe2000.warships_java.direct.general.ServerStatusRequest;
import org.junit.Test;

import java.io.FileInputStream;
import java.io.IOException;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.Properties;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

public class GeneralTest {

    private String apiKey;

    private final String instanceName = "TEST";

    public GeneralTest() throws IOException {
        Properties PROPERTIES = new Properties();
        PROPERTIES.load(new FileInputStream("Warships.properties"));
        apiKey = PROPERTIES.getProperty("APIKEY");
    }

    @Test
    public void testRateLimiter() {
        int accountId = 537376379;
        ApiBuilder.createInstance(apiKey, instanceName);
        int instanceSize = ApiBuilder.getInstanceSize();
        PlayersPersonalDataFullRequest request = PlayersPersonalDataFullRequest.createRequest().region(Region.EU).addAccountId(accountId);
        ExecutorService service = Executors.newCachedThreadPool();
        long start = System.currentTimeMillis();
        int requests = 100;
        for (int i = 0; i < requests; i++) {
            service.execute(() -> {
                PlayersPersonalDataFull result = request.fetch();
                if (result == null) {
                    System.out.println("Result was null");
                }
                assert result != null : "Result was null";
                assert result.getStatus().get() : result.getError().getMessage();
            });
        }
        service.shutdown();
        try {
            service.awaitTermination(20, TimeUnit.SECONDS);
        } catch (Exception e) {
            e.printStackTrace();
        }
        double time = (double) (System.currentTimeMillis() - start) / 1000;
        assert time <= (((double) requests / 10) * 1.3) + 1 : time;
        System.out.println("Time passed: " + time + "s");
        assert ApiBuilder.getInstanceSize() == instanceSize : ApiBuilder.getInstanceSize() + ", expected size of " + instanceSize;
    }

    @Test
    public void testShutdown() {
        ApiBuilder.createInstance(apiKey);
        assert ApiBuilder.getApiKeyAsParam(null) != null : "Api key is null";
        ApiBuilder.shutdown();
        assert ApiBuilder.getInstanceSize() == 0 : "Instance list not cleared after shutdown";
        ApiBuilder.createInstance(apiKey, "NEWAPP");
        assert ApiBuilder.getApiKeyAsParam(null) != null : "Api key is null";
        ApiBuilder.shutdown();
    }

    @Test
    public void testServerStatus() {
        ApiBuilder.createInstanceIfNoneExists(apiKey);
        ServerStatusRequest request = ServerStatusRequest.createRequest().region(Region.EU);
        ServerStatus response = request.fetch();
        assert response.getStatus().get() : "Invalid response state\n" + response.getError().toString();
        System.out.println(response);
    }
}

File: src/test/kotlin/GeneralTest.kt
import de.floribe2000.warships_java.direct.account.PlayersPersonalDataFullRequest
import de.floribe2000.warships_java.direct.api.ApiBuilder.Companion.getApiKeyAsParam
import de.floribe2000.warships_java.direct.api.ApiBuilder.Companion.getInstanceSize
import de.floribe2000.warships_java.direct.api.ApiBuilder.Companion.shutdown
import de.floribe2000.warships_java.direct.api.typeDefinitions.Region
import de.floribe2000.warships_java.direct.general.ServerStatusRequest
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertTimeout
import utilities.ITestClass
import java.io.FileInputStream
import java.time.Duration
import java.util.*
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicInteger

class GeneralTest : ITestClass {
    override val instanceName = "TEST"

    @Test
    fun testRateLimiter() {
        val accountId = 537376379
        setupApi()
        val instanceSize = getInstanceSize()
        val request = PlayersPersonalDataFullRequest().region(Region.EU).addAccountId(accountId.toLong())
        val service = Executors.newCachedThreadPool()
        val requests = 100
        val errorCount = AtomicInteger(0)

        assertTimeout(Duration.ofSeconds((requests.toDouble() / 10 * 1.5).toLong() + 1)) {
            for (i in 0 until requests) {
                service.execute {
                    val result = try {
                        request.fetch()
                    } catch (e: IllegalStateException) {
                        errorCount.addAndGet(1)
                        throw AssertionError("Request was unable to be processed.")
                    }
                    assert(result.status.get()) { result.error?.message ?: "Unable to read error message." }
                }
            }
            service.shutdown()
            try {
                service.awaitTermination(100, TimeUnit.SECONDS)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
        assert(errorCount.get() == 0) { "There were errors during the test." }
        assert(getInstanceSize() == instanceSize) { getInstanceSize().toString() + ", expected size of " + instanceSize }
    }

    @Test
    fun testShutdown() {
        setupApi()
        assert(getApiKeyAsParam(null) != null) { "Api key is null" }
        shutdown()
        assert(getInstanceSize() == 0) { "Instance list not cleared after shutdown" }
        setupApi(instanceName = "NEWAPP")
        assert(getApiKeyAsParam(null) != null) { "Api key is null" }
        shutdown()
    }

    @Test
    fun testServerStatus() {
        setupApi()
        val request = ServerStatusRequest.createRequest().region(Region.EU)
        val response = request.fetch()
        assert(response.status.get()) {
            "Invalid response state:\n ${response.error.toString()}"
        }
    }
}